Use angular-file-saver for the download files

Itxaka opened the pull request:
As pointed by @skazi0 angular-file-saver makes the downloading stuff pretty easy and it works cross-browser, fixing https://github.com/crowbar/crowbar-ui/issues/32 and probably any other issue with IE browsers.

This is only a run test, tests doesn't pass and there is no new tests done for the different functionality but its so @santiph can weight on what he thinks about this and if we should go forward with this or not.

Commit:
WIP: Test angular-file-saver for the download files

skazi0 commented on assets/app/features/upgrade/controllers/upgrade-backup.controller.js:
```diff
@@ -68,18 +75,9 @@ function() {
 
-                            backupBlob = new Blob([response.data], {type: contentType}),
-                            backupObjectUrl = URL.createObjectURL(backupBlob),
+                            backupBlob = new Blob([response.data], {type: contentType});
```
Depending on the portability requirements (i.e. how old browsers need to be supported), we can make it even simpler: if you set `responseType: 'blob'` in `crowbar-backup.factory.js` you will get `response.data` already in blob format and you can drop the whole manual `Blob` object creation.

`Blob` has pretty decent support in modern desktop browsers: https://developer.mozilla.org/en-US/docs/Web/API/Blob#Browser_compatibility

Itxaka replied:
I thougth that to trigger the download of zip files and binary files in general you were forced to set the `arraybuffer` responseType?

skazi0 replied:
Hmmm... haven't heard about such requirement. I tested `responseType: blob` in current FF and Chrome and it works correctly.

Commit:
Use Blob directly on the response
Remove uneeded stuff now that we use Blob.
Remove uneeded tests now that we dont click on an element
Add tests for the FileSaver

Itxaka commented:
@skazi0 ready for a second round of checks, made the tests work with the new fileSaver, used blob as you suggested and removed non used code.

skazi0 commented on assets/app/features/upgrade/controllers/upgrade-backup.controller.js:
```diff
@@ -12,9 +12,16 @@ function() {
+        'crowbarBackupFactory',
+        '$document',
+        'FileSaver',
+        'Blob'
```
is this still needed?

skazi0 commented on assets/app/features/upgrade/controllers/upgrade-backup.controller.spec.js:
```diff
@@ -161,21 +148,15 @@ describe('Upgrade Flow - Backup Controller', function() {
                     it('uses the default name for the file', function () {
-                        expect(downloadBackupMock.download).toEqual('crowbarBackup');
+                        expect(FileSaver.saveAs).toHaveBeenCalledWith(
+                          mockedDownloadResponse.data, 'crowbarBackup'
```
Since this test is about the filename, I'd put jasmine.anything() as the first argument. The data can be a separate test.

skazi0 commented on assets/app/features/upgrade/controllers/upgrade-backup.controller.spec.js:
```diff
@@ -210,7 +186,9 @@ describe('Upgrade Flow - Backup Controller', function() {
                     it('does not use the default name for the file', function () {
-                        expect(downloadBackupMock.download).not.toEqual('crowbarBackup');
+                        expect(FileSaver.saveAs).not.toHaveBeenCalledWith(
+                          mockedDownloadResponse.data, 'crowbarBackup'
```
You can use this place to do some basic blackbox testing of `extractFilename()`, i.e. expect the actual filename from headers here.

skazi0 reviewed (commented):
small nits in the tests. besides that, looks good.

skazi0 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(1) S50P12 (S49P13) #8 Backup Admin Server: Service Integration](https://trello.com/c/hw0n8t1G/129-1-s50p12-s49p13-8-backup-admin-server-service-integration)

skazi0 committed:
Test fixes.

skazi0 commented:
@rsalevsky this adds an npm dependency: https://github.com/crowbar/crowbar-ui/pull/82/files#diff-0a08a7565aba4405282251491979bb6bR27 please confirm it's ok from the packaging point of view

santiph commented on assets/app/features/upgrade/controllers/upgrade-backup.controller.spec.js:
```diff
@@ -161,21 +149,15 @@ describe('Upgrade Flow - Backup Controller', function() {
                     it('uses the default name for the file', function () {
-                        expect(downloadBackupMock.download).toEqual('crowbarBackup');
+                        expect(FileSaver.saveAs).toHaveBeenCalledWith(
+                          jasmine.anything(), 'crowbarBackup'
```
4 spaces please. Hasn't this been caught by travis?

santiph replied:
Also, could `anything()` be more specific?

santiph commented on assets/app/features/upgrade/controllers/upgrade-backup.controller.spec.js:
```diff
@@ -192,11 +174,6 @@ describe('Upgrade Flow - Backup Controller', function() {
                         $rootScope.$digest();
                     });
 
-                    it('click() method should have been triggered to download the backup', function () {
```
Why removed?

skazi0 replied:
This is the whole beauty of using library instead of self-crafted direct solution: you don't need to test the internals of the library, just make sure that library functions are called (`expect(FileSaver.saveAs).toHaveBeenCalledWith(...)`)

santiph replied:
You're right. I see it now

santiph commented on assets/app/features/upgrade/controllers/upgrade-backup.controller.spec.js:
```diff
@@ -209,8 +186,16 @@ describe('Upgrade Flow - Backup Controller', function() {
+
+                    it('calls saveAs with data received from the service', function () {
+                        expect(FileSaver.saveAs).toHaveBeenCalledWith(
+                          mockedDownloadResponse.data, jasmine.anything()
```
could this `anything()` be more specific?

santiph reviewed (changes requested):
Please, also include this dependency in the core module, so it can be reused on future projects
`assets/app/core/crowbar-core.module.js`

rsalevsky commented:
I have to update the package after merging this. So I would suggest once @santiph approved I will merge it and update the package.

skazi0 committed:
Added 'ngFileSaver' dependency to core module

Itxaka commented:
Thank you @skazi0 for updating the PR! :+1: 

@santiph can we merge this?

santiph commented:
@Itxaka Looks good so far. Please rebase it and I'll do a final review.

Itxaka commented:
rebased!

Commit:
Change files and tests to be in sync with latests changes from master

 - backupresponse is no longer an arraybuffer but a blob
 - $document is not used on teh backup tests anymore
 - crowbarBackupFactory.get is now crowbarUtilsFactory.getAdminBackup

Itxaka commented:
Fixes issues that came from rebasing. Tested on both firefox+chrome. Looking good!
